In [1]:
from keras.applications.inception_resnet_v2 import InceptionResNetV2
from keras.models import load_model, Sequential, Model
from keras.layers import Input, Dense
from keras.optimizers import Adam
from keras.callbacks import TensorBoard, ModelCheckpoint
from keras.preprocessing import image
from keras.applications.inception_resnet_v2 import preprocess_input, decode_predictions

import numpy as np
from keras.utils.np_utils import to_categorical
import glob
import os

from sgdr import SGDRScheduler

Using TensorFlow backend.


In [2]:
categories = ['0_buildings', '1_forest', '2_glacier', '3_mountain', '4_sea', '5_street']

In [3]:
train_datagen = image.ImageDataGenerator(rotation_range=10,
                                         width_shift_range=0.25,
                                         height_shift_range=0.25,
                                         shear_range=0.01,
                                         zoom_range=[0.75, 1.25],
                                         horizontal_flip=True,
                                         vertical_flip=False,
                                         data_format='channels_last',
                                         fill_mode='reflect',
                                         brightness_range=[0.75, 1.25]
                                        )

In [15]:
train_generator = train_datagen.flow_from_directory(
    directory='C:/Datasets/scene-classification/data_augment/train',
    target_size=(299, 299),
    color_mode='rgb',
    batch_size=4,
    class_mode='categorical',
    shuffle=True,
    seed=42
)

Found 17034 images belonging to 6 classes.


In [14]:
input_tensor = Input(shape=(299, 299, 3))
base_model = InceptionResNetV2(include_top=False, weights='imagenet', input_tensor=input_tensor, input_shape=(299, 299, 3), pooling='avg', classes=1000)
for layer in base_model.layers[:-13]:
    layer.trainable = False
    
t = base_model(input_tensor)
output_tensor = Dense(len(categories), activation='softmax')(t)
model = Model(inputs=input_tensor, outputs=output_tensor)
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_2 (InputLayer)         (None, 299, 299, 3)       0         
_________________________________________________________________
inception_resnet_v2 (Model)  (None, 1536)              54336736  
_________________________________________________________________
dense_2 (Dense)              (None, 6)                 9222      
Total params: 54,345,958
Trainable params: 4,711,398
Non-trainable params: 49,634,560
_________________________________________________________________


In [8]:
file_path = 'C:/Datasets/scene-classification/models/'
model_name = 'irnv2_adam_sgdr_v0.2-batch_size-32.hdf5'
checkpoint = ModelCheckpoint(file_path + model_name, monitor='categorical_accuracy', save_best_only=True, verbose=1, mode='max')
optimizer = Adam(lr=3e-4)
tboard = TensorBoard(log_dir='C:/Datasets/scene-classification/logs/')

In [9]:
model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['categorical_accuracy'])

In [10]:
STEP_SIZE_TRAIN = train_generator.n // train_generator.batch_size

In [11]:
STEP_SIZE_TRAIN

532

In [12]:
model.fit_generator(generator=train_generator,
                    steps_per_epoch=STEP_SIZE_TRAIN,
                    callbacks=[checkpoint, tboard],
                    epochs=1,
                    verbose=1
                   )

Epoch 1/1
532/532 [==============================] - 401s 755ms/step - loss: 0.3916 - categorical_accuracy: 0.8579

Epoch 00001: categorical_accuracy improved from -inf to 0.85784, saving model to C:/Datasets/scene-classification/models/irnv2_adam_sgdr_v0.2-batch_size-32.hdf5


In [6]:
file_path = 'C:/Datasets/scene-classification/models/'
model_name = 'irnv2_adam_sgdr_v0.2-batch_size-16.hdf5'
checkpoint = ModelCheckpoint(file_path + model_name, monitor='categorical_accuracy', save_best_only=True, verbose=1, mode='max')
optimizer = Adam(lr=3e-4)
tboard = TensorBoard(log_dir='C:/Datasets/scene-classification/logs/')

In [7]:
model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['categorical_accuracy'])

In [8]:
STEP_SIZE_TRAIN = train_generator.n // train_generator.batch_size
STEP_SIZE_TRAIN

1064

In [9]:
model.fit_generator(generator=train_generator,
                    steps_per_epoch=STEP_SIZE_TRAIN,
                    callbacks=[checkpoint, tboard],
                    epochs=1,
                    verbose=1
                   )

Epoch 1/1
1064/1064 [==============================] - 401s 377ms/step - loss: 0.4244 - categorical_accuracy: 0.8430

Epoch 00001: categorical_accuracy improved from -inf to 0.84299, saving model to C:/Datasets/scene-classification/models/irnv2_adam_sgdr_v0.2-batch_size-16.hdf5


In [10]:
file_path = 'C:/Datasets/scene-classification/models/'
model_name = 'irnv2_adam_sgdr_v0.2-batch_size-8.hdf5'
checkpoint = ModelCheckpoint(file_path + model_name, monitor='categorical_accuracy', save_best_only=True, verbose=1, mode='max')
optimizer = Adam(lr=3e-4)
tboard = TensorBoard(log_dir='C:/Datasets/scene-classification/logs/')

In [11]:
model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['categorical_accuracy'])

In [12]:
STEP_SIZE_TRAIN = train_generator.n // train_generator.batch_size
STEP_SIZE_TRAIN

2129

In [13]:
model.fit_generator(generator=train_generator,
                    steps_per_epoch=STEP_SIZE_TRAIN,
                    callbacks=[checkpoint, tboard],
                    epochs=1,
                    verbose=1
                   )

Epoch 1/1
2129/2129 [==============================] - 443s 208ms/step - loss: 0.4835 - categorical_accuracy: 0.8198

Epoch 00001: categorical_accuracy improved from -inf to 0.81992, saving model to C:/Datasets/scene-classification/models/irnv2_adam_sgdr_v0.2-batch_size-8.hdf5


In [16]:
file_path = 'C:/Datasets/scene-classification/models/'
model_name = 'irnv2_adam_sgdr_v0.2-batch_size-4.hdf5'
checkpoint = ModelCheckpoint(file_path + model_name, monitor='categorical_accuracy', save_best_only=True, verbose=1, mode='max')
optimizer = Adam(lr=3e-4)
tboard = TensorBoard(log_dir='C:/Datasets/scene-classification/logs/')

In [17]:
model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['categorical_accuracy'])

In [18]:
STEP_SIZE_TRAIN = train_generator.n // train_generator.batch_size
STEP_SIZE_TRAIN

4258

In [19]:
model.fit_generator(generator=train_generator,
                    steps_per_epoch=STEP_SIZE_TRAIN,
                    callbacks=[checkpoint, tboard],
                    epochs=1,
                    verbose=1
                   )

Epoch 1/1
4258/4258 [==============================] - 528s 124ms/step - loss: 0.5940 - categorical_accuracy: 0.7846

Epoch 00001: categorical_accuracy improved from -inf to 0.78462, saving model to C:/Datasets/scene-classification/models/irnv2_adam_sgdr_v0.2-batch_size-4.hdf5


__Batch size = 32__
<br>
loss = 0.3916, categorical_accuracy = 0.8579
<br>
<br>
__Batch size = 16__
<br>
loss = 0.4244, categorical_accuracy = 0.84229
<br>
<br>
__Batch size = 8__
<br>
loss = 0.4835, categorical_accuracy = 0.8198
<br>
<br>
__Batch size = 4__
<br>
loss = 0.5940, categorical_accuracy = 0.7846
<br>